### PR TITLE
IoT Perf changes

### PR DESCRIPTION
**Pull request by CIPop**
* Fixing writer semaphore. 
* Limiting output size to 2GB. 
* Checking method result.

### CODE REVIEW

File: e2e/stress/IoTClientPerf/Scenarios/ServiceClientScenarios.cs [hunk 1]
```diff
@@ -102,11 +103,17 @@ protected async Task CallMethodAsync(CancellationToken ct)
             try
             {
                 var methodCall = new CloudToDeviceMethod(TestMethodName);
-                Task t = s_sc.InvokeDeviceMethodAsync(Configuration.Stress.GetDeviceNameById(_id, _authType), methodCall);
+                Task<CloudToDeviceMethodResult> t = s_sc.InvokeDeviceMethodAsync(Configuration.Stress.GetDeviceNameById(_id, _authType), methodCall);
                 _mMethod.ScheduleTime = _swMethod.ElapsedMilliseconds;
 
                 _swMethod.Restart();
-                await t.ConfigureAwait(false);
+                CloudToDeviceMethodResult result = await t.ConfigureAwait(false);
```
Comment (davilu) [inline e2e/stress/IoTClientPerf/Scenarios/ServiceClientScenarios.cs]: Please use `finally`.
```
try
{
  ...
  if (result.Status != MethodPassStatus) 
  {
     _mMethod.ErrorMessage = $"Method Failed - Status: {result.Status} Payload:{result.GetPayloadAsJson()}";
  }
}
catch(Exception ex)
{
  _mMethod.ErrorMessage = $"{ex.GetType().Name} - {ex.Message}";
  throw;
}
finally
{
  _mMethod.ExecuteTime = _swMethod.ElapsedMilliseconds;
   await _writer.WriteAsync(_mMethod).ConfigureAwait(false);
}
```

File: e2e/stress/IoTClientPerf/Reporting/ResultWriterFile.cs [hunk 2]
```diff
@@ -11,44 +12,96 @@ namespace Microsoft.Azure.Devices.E2ETests
...
+                    if (_header != null)
+                    {
+                        await _writer.WriteLineAsync(_header).ConfigureAwait(false);
+                        _fileSize += _header.Length;
+                    }
+                }
+
+                await _writer.WriteLineAsync(s).ConfigureAwait(false);
+
+                if (++_bufferedLines > MaximumLinesBuffer)
```
[the remark this answers is not in the document]
Comment (CIPop) [reply e2e/stress/IoTClientPerf/Reporting/ResultWriterFile.cs]: Makes sense: removing.

File: e2e/stress/IoTClientPerf/Reporting/ResultWriterFile.cs [hunk 3]
```diff
@@ -11,44 +12,96 @@ namespace Microsoft.Azure.Devices.E2ETests
...
+
+            string dir = Path.GetDirectoryName(_fileName);
+            string file = Path.GetFileNameWithoutExtension(_fileName);
+            string ext = Path.GetExtension(_fileName);
 
-            if (++_bufferedLines > MaximumLinesBuffer)
+            return Path.Combine(dir, $"{file}_{_fileCount}{ext}");
+        }
+
+        private async Task CreateNewLogFileAsync()
```
[the remark this answers is not in the document]
Comment (CIPop) [reply e2e/stress/IoTClientPerf/Reporting/ResultWriterFile.cs]: Renamed to RotateLogFileAsync - let me know if you have any suggestions here.
I simplified by removing InitWriter as well.

File: e2e/stress/IoTClientPerf/Scenarios/ServiceClientScenarios.cs (hunk 1 above)
Comment (CIPop) [reply e2e/stress/IoTClientPerf/Scenarios/ServiceClientScenarios.cs]: Corrected here and in DeviceScenarios

**Comment by CIPop** [comment]
@davilu I've pushed another commit for the PR comments. PTAL

File: e2e/stress/IoTClientPerf/Scenarios/ServiceClientScenarios.cs [hunk 4]
```diff
@@ -101,22 +111,33 @@ protected async Task CallMethodAsync(CancellationToken ct)
...
+                // Check method result.
+                if (result.Status != MethodPassStatus)
+                {
+                    throw new InvalidOperationException($"Method Failed status={result.Status} Payload:{result.GetPayloadAsJson()}");
+                }
             }
             catch (Exception ex)
             {
                 _mMethod.ErrorMessage = $"{ex.GetType().Name} - {ex.Message}";
                 exInfo = ExceptionDispatchInfo.Capture(ex);
```
Comment (davilu) [inline e2e/stress/IoTClientPerf/Scenarios/ServiceClientScenarios.cs]: just throw. No need to capture.

Comment (CIPop) [reply e2e/stress/IoTClientPerf/Scenarios/ServiceClientScenarios.cs]: That won't work: I need to capture so that I can write the exception to the log.

Comment (CIPop) [reply e2e/stress/IoTClientPerf/Scenarios/ServiceClientScenarios.cs]: Discussed offline: it's now in the finally block so it should always execute - no need to capture.

File: e2e/stress/IoTClientPerf/Scenarios/ServiceClientScenarios.cs (hunk 1 above)
Comment (CIPop) [reply e2e/stress/IoTClientPerf/Scenarios/ServiceClientScenarios.cs]: Remove `InvalidOperationException` that is thrown and caught. Replace message with the same pattern but with a non-exception for error title.

Comment (CIPop) [reply e2e/stress/IoTClientPerf/Scenarios/ServiceClientScenarios.cs]: @davilu Can't remove the exception: the aggregate reports are based on the Task completing, failing or getting cancelled.

Changed all the other items.

**Pull request merged**